Remove `feature` from the list of well known check-cfg name

Urgau opened the pull request:
This PR removes `feature` from the list of well known check-cfg.

This is done for multiple reasons:
 - Cargo is the source of truth, rustc shouldn't have any knowledge of it
 - It creates a conflict between Cargo and rustc when there are no features defined.
   In this case Cargo won't pass any `--check-cfg` for `feature` since no feature will ever be passed, but rustc by having in it's list adds a implicit `cfg(feature, values(any()))` which is completely wrong. Having any cfg `feature` is unexpected not allow any `feature` value.

While doing this, I took the opportunity to specialise the diagnostic a bit for the case above. 

r? @petrochenkov

Urgau committed:
Remove `feature` from the list of well known check-cfg name

Urgau committed:
Improve diagnostic for unexpected feature config name from Cargo

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6922879396/job/18830118537) [(plain)](https://github.com/rust-lang/rust/commit/e1e1cb47a22c86a943bbd71d6ba11d1bf1557495/checks/18830118537/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=Urgau
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_51285513-cc01-47f3-8804-f77f7868f1d7
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=c1e74cfce5cd41800832fc13b03953cc30021299
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_51285513-cc01-47f3-8804-f77f7868f1d7
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_51285513-cc01-47f3-8804-f77f7868f1d7
GITHUB_TRIGGERING_ACTOR=Urgau
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/118071/merge
GITHUB_WORKFLOW_SHA=c1e74cfce5cd41800832fc13b03953cc30021299
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
   Compiling memchr v2.5.0
   Compiling std v0.0.0 (/checkout/library/std)
error: unexpected `feature` as condition name
  |
  = help: was set with `--cfg` but isn't in the `--check-cfg` expected names
  = note: `-D unexpected-cfgs` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(unexpected_cfgs)]`
error: could not compile `std` (build script) due to previous error
warning: build failed, waiting for other jobs to finish...
fatal error: failed to build sysroot: sysroot build failed
Build completed unsuccessfully in 0:01:29
```

</details>

Urgau committed:
[Miri] Do not respect RUSTC_HOST_FLAGS since RUSTFLAGS isn't

When building the standard library with Miri, it appears that cargo-miri
does not respect the RUSTFLAGS and even if they did it would be wrong
since they are created for ToolRustc not Std. To avoid errors ignore
RUSTC_HOST_FLAGS for Miri.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6923265971/job/18830958052) [(plain)](https://github.com/rust-lang/rust/commit/5b41facf4d0c144966d1c355125a686184dc89e9/checks/18830958052/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=Urgau
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_42288264-99ce-4ea8-a207-d7d8e2900c35
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=fd4f4ce1714362f08200806dd3fc27eb46143b23
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_42288264-99ce-4ea8-a207-d7d8e2900c35
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_42288264-99ce-4ea8-a207-d7d8e2900c35
GITHUB_TRIGGERING_ACTOR=Urgau
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/118071/merge
GITHUB_WORKFLOW_SHA=fd4f4ce1714362f08200806dd3fc27eb46143b23
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
   Compiling memchr v2.5.0
   Compiling std v0.0.0 (/checkout/library/std)
error: unexpected `feature` as condition name
  |
  = help: was set with `--cfg` but isn't in the `--check-cfg` expected names
  = note: `-D unexpected-cfgs` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(unexpected_cfgs)]`
error: could not compile `std` (build script) due to previous error
warning: build failed, waiting for other jobs to finish...
fatal error: failed to build sysroot: sysroot build failed
Build completed unsuccessfully in 0:01:43
```

</details>

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 3fda3efe2072cd7ae7cb82048b31d80f9f0fa0c0 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 3fda3efe2072cd7ae7cb82048b31d80f9f0fa0c0 -->
<!-- homu: {"type":"Approved","sha":"3fda3efe2072cd7ae7cb82048b31d80f9f0fa0c0","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

petrochenkov commented:
I don't understand full implications of the Miri change though.
If it's not strictly necessary for this PR, then I'd suggest landing it separately with a review from qualified people.

Urgau commented:
> I don't understand full implications of the Miri change though. If it's not strictly necessary for this PR, then I'd suggest landing it separately with a review from qualified people.

It's unfortunately required, and the reason it is, is because of Cargo not passing `RUSTFLAGS` to proc-macro made the `bootstrap` tool create the `RUSTC_HOST_FLAGS` which force pass some flags to our `rustc` shim. Inside that environment variable are some `--check-cfg` flags that activates checking of cfgs but the way the standard library is build by Miri makes `cargo-miri` not use `bootstrap` and so neither `-Zcheck-cfg` nor our extra `--check-cfg` are passed making the build fail on all but the `bootstrap` cfg.

Ignoring `RUSTC_HOST_FLAGS` when we detect Miri is the easiest way I found to solve the issue.

Urgau commented:
@rustbot ready

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 54c122ec0d45799e795f8b2aca38227ce9b05fa5 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 54c122ec0d45799e795f8b2aca38227ce9b05fa5 -->
<!-- homu: {"type":"Approved","sha":"54c122ec0d45799e795f8b2aca38227ce9b05fa5","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 54c122ec0d45799e795f8b2aca38227ce9b05fa5 with merge 855c6836b7c476385db95a0a95f2e57f5ff32fe1...
<!-- homu: {"type":"BuildStarted","head_sha":"54c122ec0d45799e795f8b2aca38227ce9b05fa5","merge_sha":"855c6836b7c476385db95a0a95f2e57f5ff32fe1"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6954188207/job/18923864361)
Approved by: petrochenkov
Pushing 855c6836b7c476385db95a0a95f2e57f5ff32fe1 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6954188207/job/18923864361"},"merge_sha":"855c6836b7c476385db95a0a95f2e57f5ff32fe1"} -->

rust-timer commented:
Finished benchmarking commit (855c6836b7c476385db95a0a95f2e57f5ff32fe1): [comparison URL](https://perf.rust-lang.org/compare.html?start=cc4bb0de204516c2852f70df0f3849e542e2c95e&end=855c6836b7c476385db95a0a95f2e57f5ff32fe1&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=cc4bb0de204516c2852f70df0f3849e542e2c95e&end=855c6836b7c476385db95a0a95f2e57f5ff32fe1&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=cc4bb0de204516c2852f70df0f3849e542e2c95e&end=855c6836b7c476385db95a0a95f2e57f5ff32fe1&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.8% | [-0.8%, -0.8%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=cc4bb0de204516c2852f70df0f3849e542e2c95e&end=855c6836b7c476385db95a0a95f2e57f5ff32fe1&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range         | count |
|:----------------------------------:|:----:|:-------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -             | 0     |
| Regressions ❌ <br /> (secondary)  | 9.7% | [2.6%, 25.6%] | 6     |
| Improvements ✅ <br /> (primary)   | -    | -             | 0     |
| Improvements ✅ <br /> (secondary) | -    | -             | 0     |
| All ❌✅ (primary)                 | -    | -             | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=cc4bb0de204516c2852f70df0f3849e542e2c95e&end=855c6836b7c476385db95a0a95f2e57f5ff32fe1&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 676.862s -> 677.73s (0.13%)
**Artifact size**: 313.76 MiB -> 313.77 MiB (0.01%)
